### ISSUE TITLE
Inconsistent metric family meta data requirements

### ISSUE THREAD

**Issue by ColinH** [OpenObservability/OpenMetrics#215]
In the data model section it is stated that

> A MetricFamily MUST have a name, HELP, TYPE, and UNIT metadata.

however the section on metric family meta data, and other parts of the document, imply that HELP, TYPE and UNIT are optional.

**Comment by brian-brazil** [comment]
This is data model vs how it's represented on the wire. Metrics always have these, but they may not explicitly appear.

**Comment by ColinH** [comment]
Is the data model that of Prometheus or that of the OpenMetrics exposition format?

**Comment by brian-brazil** [comment]
Is is the data model of OpenMetrics.

**Comment by ColinH** [comment]
What are the implications of this requirement when the text format makes all of these meta data fields optional? Would it make any difference if it were removed?

**Comment by brian-brazil** [comment]
There are no implications I see, it's just how the underlying data structure is worded.

**Comment by ColinH** [comment]
So the data model says that help, type and unit are mandatory, but in both of the encodings supposedly adhering to said data model these fields are optional ... why does the data model include this?

If the requirement has no implications then it's not a requirement, it's something that should be removed from the document. At best it serves no purpose, at worst it's misleading or a distraction.

**Comment by brian-brazil** [comment]
The wording is a bit weird due to RFC2119. All it's saying is that metrics have fields called help, type and unit. Without that the data model would be missing key fields.

How those are parsed/exposed is orthogonal.

**Comment by d-frey** [comment]
What I find confusing is the lack of a clear distinction between a value that is missing and assigning meaning to special values. It seems that the data model says the values are mandatory, but the empty string has special meaning in the representation. Specifically: A unit is mandatory in the data model, but the empty string as a unit can be left out in the representation. Also working with SQL, I think that not having a unit value is equivalent to SQL's NULL, while an empty string is a value and thus equivalent to "". Mixing both in the data model and the representation does create a clear picture for me.

In summary, I think that using the SQL wording again, the data model should say that there is a column "unit", but it is nullable. If it is *not* NULL, it should be a string that is *not* empty.

**Comment by brian-brazil** [comment]
I think it best not to mix in additional concepts and terminology that not all readers may not be familiar with, and the handling of empty unit fields are already explained without having to resort to nullability. Such notions would likely only cause confusion, as it would seem that null and empty are different - which they wouldn't be.

> In summary, I think that using the SQL wording again, the data model should say that there is a column "unit", but it is nullable.

This would be incorrect, unit is not nullable. It is a string, of possibly zero length.

**Comment by d-frey** [comment]
I just used SQL terminology to explain my point, not suggesting you should use it. But the concept behind this is important, as you have a unit which can be an empty string. This has a special meaning and special handling, e.g. if it is non-empty, it must be a suffix of the name (plus the additional underscore). So what is that special meaning? It means: "There is no actual unit". This would IMHO be better expressed by saying that the unit is optional. And if a unit is given, it must be non-empty.

**Comment by brian-brazil** [comment]
You're missing a subtlety there though, which is what this bit of the spec is capturing. Even if the unit is empty, there is still a unit. Thus data model wise the unit is not optional.

**Comment by ColinH** [comment]
> The wording is a bit weird due to RFC2119.

In this case it would be sufficient to change

> A MetricFamily MUST have a name, HELP, TYPE, and UNIT metadata.

to

> A MetricFamily MUST have a name, and MAY have HELP, TYPE, and UNIT metadata.

but I will now drop this subject.

Btw, regarding the use of RFC2119 words, if one were to write "a MetricFamily is identified by its name" instead of "a MetricFamily MUST have a  name" then it would automatically imply both that (a) it has a name, and (b) the name is unique, without requiring the second clause for (b).

This is in line with how, In general, I find that RFC2119 words often work better for restrictions and similar rather than for basic definitions. Instead of "a MetricSet MUST consist of MetricFamilies and MAY be empty." it feels more natural to write "a MetricSet is a (possibly empty) collection of MetricFamilies".

Compare this to something like "the combined length of the label names and values of an Exemplar's LabelSet MUST NOT exceed 128 UTF-8 characters." where the use of "MUST NOT" is clearly appropriate and one would be hard pressed to find a better formulation without it.

**Comment by d-frey** [comment]
I think you are missing my point, but like Colin, I give up.
